### PR TITLE
Feature/api

### PR DESCRIPTION
**Pull request by FredrikAugust**
- [x] Generalize setup for tests

## Mitochondria implementation
- [x] `transaction`
  - [x] test
- [x] `recurring`
  - [ ] test
- [x] `balance`
  - [x] test
- [x] `month`
  - [x] test
- [x] `company`
  - [x] test
- [x] `user`
  - [x] test

----------------------------------------

Big PR, so here's some things that you might find confusing

### `run.test.ts`

Why is this here? And why don't you just run the test files in parallel?

The point here is that the file is set up with `describe`s, which run synchronously, so by doing this, I can make sure that the backend tests don't run in parallel, which is necessary to avoid conflicts in the tests.

### Creating a transaction before each transaction test

I had to do this to avoid issues with async locks messing with the tests. I have currently not found a workaround which performs better, but it is something I plan to look into.

### `yarn test:api`

This will run the tests that require a connection to the backend as well. 

### `yarn test:all`

Run all tests. Requires a backend to work.

### Reducer actions?

I am going to hold off on implementing the reducer actions here, as that will be another 1000 LOC. Will branch out from this PR and implement it there.

### Recurring transaction tests?

This is currently being discussed [here](https://github.com/Kodeworks/liquidator-backend/issues/48) and [here](https://github.com/Kodeworks/liquidator-backend/issues/52).

### CODE REVIEW

**Comment by FredrikAugust** [comment]
Will setup the backend on heroku or something and write tests against that, because mocking everything with nock will take ages. I'll prefix the tests with `.backend` or something so you can choose not to run them when working locally, but they should definitely be in place. 

Like this @NicolajNiels1, you won't have to wait a long time for the tests to run, and I get to test the backend and frontend in one!

File: src/mitochondria/recurring.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    `?company_id=${companyId}&id=${id}`,
+    {
+      method: 'DELETE',
+    }
+  );
+
+export const getAllRecurringTransactions = async (
+  companyId: number,
+  offset: number = 0,
+  limit?: number
```
Comment (FredrikAugust) [inline src/mitochondria/recurring.ts]: change to default

File: src/mitochondria/recurring.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+) =>
+  await fetchWithCallback<IPaginated<IRecurringTransaction>>(
+    '/recurring/all/',
+    `?company_id=${companyId}&offset=${offset}&limit=${limit}`
+  );
+
+export const getActiveRecurringTransactions = async (
+  companyId: number,
+  offset: number = 0,
+  limit?: number
```
Comment (FredrikAugust) [inline src/mitochondria/recurring.ts]: change to default

File: src/mitochondria/recurring.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  );
+
+/**
+ * @param date "ISO-8601"
+ */
+export const getRecurringTransactionsByDate = async (
+  companyId: number,
+  date: string,
+  offset: number = 0,
+  limit?: number
```
Comment (FredrikAugust) [inline src/mitochondria/recurring.ts]: change to default

File: src/mitochondria/recurring.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+/**
+ * @param startDate "Inclusive (ISO-8601)"
+ * @param endDate "Inclusive (ISO-8601)"
+ */
+export const getRecurringTransactionsByDateRange = async (
+  companyId: number,
+  startDate: string,
+  endDate: string,
+  offset: number = 0,
+  limit?: number
```
Comment (FredrikAugust) [inline src/mitochondria/recurring.ts]: change to default

File: src/mitochondria/recurring.ts (hunk 1 above)
Comment (FredrikAugust) [reply src/mitochondria/recurring.ts]:
```suggestion
  limit: number = 0
```

**Comment by NicolajNiels1** [comment]
@FredrikAugust 	@KaizerTilt 	
About, that bug we discussed in the balances files. Is that something that should be caught in the tests? 🤔

**Comment by FredrikAugust** [comment]
No. The bug was that the backend API was changed, and the frontend code was not updated to reflect that.

The fact that the API was changed from the initial "state" was caught in the test, which was the purpose of it.

**Comment by FredrikAugust** [comment]
It should however have been caught in a `Company` component test, which I do not believe exists yet. Just realised that you might not have referred to the API tests.

File: README.md [hunk 5]
```diff
@@ -30,4 +30,10 @@ yarn storybook
 
 Other than that, it's just a create-react-app, so the normal `yarn build` etc. works!
 
+To test the application, we have three different commands;
+
+1. `yarn test` -- Runs the component and reducer tests. Does not require a backend running.
+2. `yarn test:api` -- Only run the backend integration tests. Requires a running instance of [`liquidator-backend`](https://github.com/kodeworks/liquidator-backend).
```
Comment (NicolajNiels1) [inline README.md]: Make CI test for integration test with backend.

**Comment by FredrikAugust** [comment]
That requires a backend running. We'll be looking at it tomorrow.

File: README.md (hunk 5 above)
Comment (NicolajNiels1) [reply README.md]: (Not required for this PR. The above comment was just to make a new Issue)

File: src/components/pages/Balances.tsx [hunk 6]
```diff
@@ -69,27 +73,29 @@ const Balances: React.FC<IProps> = props => {
...
+          const month = await BalancesAPI.getMonth(
+            monthChosen.month() + 1,
+            monthChosen.year(),
+            currentCompany.id
           );
+          newEntries[entryKey] = createBalanceEntriesFromMonth(month);
+          setEntries(newEntries);
+        } catch (e) {
+          alert('Oopsie');
+          newEntries[entryKey] = [];
```
[the remark this answers is not in the document]
Comment (KaizerTilt) [reply src/components/pages/Balances.tsx]: Actually, moving `setEntries(newEntries)` out of the try catch altogether might be  better?

File: robohydra/plugins/balancesplugin/months.js [hunk 7]
```diff
@@ -1,54 +1,54 @@
 module.exports.june = [
```
[the remark this answers is not in the document]
Comment (FredrikAugust) [reply robohydra/plugins/balancesplugin/months.js]: Good point. Will remove it on Monday.

I think I will hold off on merging this PR until next week, as I want to reflect on some of the decisions made throughout the weekend.

File: src/components/atoms/TransactionEntry.tsx [hunk 8]
```diff
@@ -54,14 +54,15 @@ const TransactionEntry: React.FC<IProps> = props => {
     >
       <h4>{props.description}</h4>
       <strong>
-        {props.type === 'expense'
+        {props.type === 'EX'
           ? `(${(money / 100).toFixed(2)})`
```
[the remark this answers is not in the document]
Comment (FredrikAugust) [reply src/components/atoms/TransactionEntry.tsx]: Absolutely.

File: robohydra/plugins/balancesplugin/months.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (KaizerTilt) [reply robohydra/plugins/balancesplugin/months.js]: Sounds good :+1:

**Pull request merged**